# Práctica 1

### Naive-Bayes

Grupo 1461 - Pareja 3

Francisco Javier Arribas Gozalo

### Naive-Bayes propio

Se ha implementado el Clasificador Naive-Bayes en el documento ClasificadorNB.py

Empezamos probando el primer conjunto de datos, tic-tac-toe:

In [10]:
import numpy as np
from Datos import Datos
from EstrategiaPatrocinado import ValidacionSimple, ValidacionCruzada
from Clasificador import Clasificador
from ClasificadorNB import ClasificadorNaiveBayes

if __name__ == "__main__":
    # Primer set de datos
    dataset = Datos("ConjuntosDatosIntroFAA/tic-tac-toe.csv")

    # Clasificador
    NB = ClasificadorNaiveBayes() # Laplace - False
    NBL = ClasificadorNaiveBayes(laplace = True) # Laplace - True

    # EstrategiaPatrocinado
    estrategia = ValidacionSimple()
    estrategia.creaParticiones(dataset)

    estrategia2 = ValidacionCruzada()
    estrategia2.creaParticiones(dataset)

    # Naive Bayes tic tac toe
    print("Tic-tac-toe, sin Laplace")
    listaErrores, mConf = NB.validacion(estrategia, dataset, NB, seed=None)

    print("\nErrores en Naive Bayes, validacion simple: ")
    print(listaErrores[0])

    listaErrores, mConf = NB.validacion(estrategia2, dataset, NB, seed=None)
    print("\nErrores en Naive Bayes, validacion cruzada: ")
    print(listaErrores)
    print("\nMedia errores: ")
    print(np.mean (listaErrores))

    print("\n\nTic-tac-toe, con Laplace")
    listaErrores, mConf = NBL.validacion(estrategia, dataset, NBL, seed=None)

    print("\nErrores en Naive Bayes, validacion simple: ")
    print(listaErrores[0])

    listaErrores, mConf = NBL.validacion(estrategia2, dataset, NBL, seed=None)
    print("\nErrores en Naive Bayes, validacion cruzada: ")
    print(listaErrores)
    print("\nMedia errores: ")
    print(np.mean (listaErrores))

Tic-tac-toe, sin Laplace

Errores en Naive Bayes, validacion simple: 
0.005235602094240838

Errores en Naive Bayes, validacion cruzada: 
[0.0, 0.005235602094240838, 0.005235602094240838, 0.005235602094240838, 0.005154639175257732, 0.005235602094240838, 0.005235602094240838, 0.0, 0.005235602094240838, 0.0]

Media errores: 
0.003656825174070276


Tic-tac-toe, con Laplace

Errores en Naive Bayes, validacion simple: 
0.005235602094240838

Errores en Naive Bayes, validacion cruzada: 
[0.0, 0.005235602094240838, 0.005235602094240838, 0.005235602094240838, 0.005154639175257732, 0.005235602094240838, 0.005235602094240838, 0.0, 0.005235602094240838, 0.0, 0.0, 0.005235602094240838, 0.0, 0.005235602094240838, 0.0]

Media errores: 
0.0031359637286122954


Continuamos probando el segundo conjunto de datos, german:

In [14]:
import numpy as np
from Datos import Datos
from EstrategiaPatrocinado import ValidacionSimple, ValidacionCruzada
from Clasificador import Clasificador
from ClasificadorNB import ClasificadorNaiveBayes

if __name__ == "__main__":
    # Segundo set de datos
    dataset2 = Datos("ConjuntosDatosIntroFAA/german.csv")

    # Clasificador
    NB2 = ClasificadorNaiveBayes() # Laplace - False
    NB2L = ClasificadorNaiveBayes(laplace = True) # Laplace - True

    # EstrategiaPatrocinado
    estrategia3 = ValidacionSimple(20, 10)
    estrategia3.creaParticiones(dataset2)

    estrategia4 = ValidacionCruzada()
    estrategia4.creaParticiones(dataset2)

    # Naive Bayes german
    print("\n\n\nGerman, sin Laplace")
    listaErrores2, mConf = NB2.validacion(estrategia3, dataset2, NB2, seed=None)

    print("\nErrores en Naive Bayes, validacion simple: ")
    print(listaErrores2[0])

    listaErrores2, mConf = NB2.validacion(estrategia4, dataset2, NB2, seed=None)
    print("\nErrores en Naive Bayes, validacion cruzada: ")
    print(listaErrores2)
    print("\nMedia errores: ")
    print(np.mean (listaErrores2))

    print("\n\nGerman, con Laplace")
    listaErrores2, mConf = NB2L.validacion(estrategia3, dataset2, NB2L, seed=None)

    print("\nErrores en Naive Bayes, validacion simple: ")
    print(listaErrores2[0])

    listaErrores2, mConf = NB2L.validacion(estrategia4, dataset2, NB2L, seed=None)
    print("\nErrores en Naive Bayes, validacion cruzada: ")
    print(listaErrores2)
    print("\nMedia errores: ")
    print(np.mean (listaErrores2))




German, sin Laplace

Errores en Naive Bayes, validacion simple: 
0.0

Errores en Naive Bayes, validacion cruzada: 
[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.005, 0.005, 0.005, 0.005]

Media errores: 
0.002


German, con Laplace

Errores en Naive Bayes, validacion simple: 
0.0

Errores en Naive Bayes, validacion cruzada: 
[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.005, 0.005, 0.005, 0.005, 0.0, 0.0, 0.0, 0.0, 0.005]

Media errores: 
0.0016666666666666668


Tenemos los siguientes resultados:

| Dataset       | Tipo Validación        | Laplace | Errores | Media de errores |
|---------------|------------------------|---------|---------|------------------|
| Tic-tac-toe   | Simple                 | No      | 0.005235602094240838 | 0.005235602094240838
| Tic-tac-toe   | Simple                 | Sí      | 0.005235602094240838 | 0.005235602094240838
| Tic-tac-toe   | Cruzada                | No      | [0.0, 0.005235602094240838, 0.005235602094240838, 0.005235602094240838, 0.005154639175257732, 0.005235602094240838, 0.005235602094240838, 0.0, 0.005235602094240838, 0.0] | 0.003656825174070276
| Tic-tac-toe   | Cruzada                | Sí      | [0.0, 0.005235602094240838, 0.005235602094240838, 0.005235602094240838, 0.005154639175257732, 0.005235602094240838, 0.005235602094240838, 0.0, 0.005235602094240838, 0.0, 0.0, 0.005235602094240838, 0.0, 0.005235602094240838, 0.0] | 0.0031359637286122954
| German        | Simple                 | No      | 0.0 | 0.0
| German        | Simple                 | Sí      | 0.0 | 0.0
| German        | Cruzada                | No      | [0.005, 0.0, 0.0, 0.0, 0.0, 0.005, 0.0, 0.0, 0.0, 0.005] | 0.0015
| German        | Cruzada                | Sí      | [0.005, 0.0, 0.0, 0.0, 0.0, 0.005, 0.0, 0.0, 0.0, 0.005, 0.0, 0.0, 0.0, 0.005, 0.005] | 0.0016666666666666668

En general, la idea es que la validación cruzada reduce la cantidad de errores que encontramos. Este efecto es fácilmente observable en Tic-tac-toe.
No podemos verlo en German, puesto que lla cantidad de errores que está cometiendo el algoritmo es tan pequeña que no parece dar ningún error en la mayoría de ejecuciones. Con más ejecuciones y más datos, observaríamos el mismo efecto que en tic-tac-toe.

El objetivo de usar Laplace, por otro lado, es el de suavizar el algoritmo al encontrarnos probabilidad 0, sumándole una constante alpha. Esto hace que nuestro Naive-Bayes se ajuste mejor a la realidad en caso de que nuestro *dataset* no tenga en cuenta todos los casos reales, y asigne por tanto una probabilidad de 0%.

### Naive-Bayes con Scikit-Learn

Este algoritmo se ha implementado en el archivo Scikit.py

In [9]:
from abc import ABCMeta, abstractmethod
import numpy as np
from Datos import Datos

from sklearn import preprocessing
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.compose import ColumnTransformer 
from sklearn.naive_bayes import MultinomialNB
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import ShuffleSplit

# https://github.com/scikit-learn/scikit-learn/blob/36958fb24/sklearn/naive_bayes.py#L750
"""
    Naive Bayes classifier for multinomial models.
    The multinomial Naive Bayes classifier is suitable for classification with
    discrete features (e.g., word counts for text classification). The
    multinomial distribution normally requires integer feature counts. However,
    in practice, fractional counts such as tf-idf may also work.
    Read more in the :ref:`User Guide <multinomial_naive_bayes>`.
"""
dataset = Datos("ConjuntosDatosIntroFAA/german.csv")
features = dataset.nominalAtributos[:-1]

ct = ColumnTransformer(dataset.nominalAtributos[:-1])

validacionSimple = ShuffleSplit(len(dataset.datos), test_size=.2, random_state=0)
validacionCruzada = 5

# Validacion Simple, sin Laplace
atributos = preprocessing.OneHotEncoder(sparse = False)
X = atributos.fit_transform(dataset.datos[:, :-1])
Y = dataset.datos[:, -1]

clasificador = MultinomialNB(alpha = 0.1, class_prior = None, fit_prior = True) # Alpha 0 -> No Laplace
clasificador.fit(X, Y)

valSimple = cross_val_score(clasificador, X, Y, cv = validacionSimple)

print("Validacion Simple sin usar Laplace:")
print("\tMedia: " + str(1 - np.mean(valSimple)))
print("\tVarianza: " + str(np.var(valSimple)) + "\n")
print("\n")

# Validacion Simple, con Laplace
atributos = preprocessing.OneHotEncoder(sparse = False)

X = atributos.fit_transform(dataset.datos[:, :-1])
Y = dataset.datos[:, -1]

clasificador = MultinomialNB(alpha = 1, class_prior = None, fit_prior = True) # Alpha 1 -> Laplace
clasificador.fit(X, Y)

valSimple = cross_val_score(clasificador, X, Y, cv = validacionSimple)

print("Validacion Simple usando Laplace:")
print("\tMedia: " + str(1 - np.mean(valSimple)))
print("\tVarianza: " + str(np.var(valSimple)) + "\n")
print("\n")

# Validacion Cruzada, sin Laplace
atributos = preprocessing.OneHotEncoder(sparse = False)
X = atributos.fit_transform(dataset.datos[:, :-1])
Y = dataset.datos[:, -1]

clasificador = MultinomialNB(alpha = 0.1, class_prior = None, fit_prior = True) # Alpha 0 -> No Laplace
clasificador.fit(X, Y)

valCruzada = cross_val_score(clasificador, X, Y, cv = validacionCruzada)

print("Validacion Cruzada sin usar Laplace:")
print("\tMedia: " + str(1 - np.mean(valCruzada)))
print("\tVarianza: " + str(np.var(valCruzada)) + "\n")
print("\n")

# Validacion Cruzada, con Laplace
atributos = preprocessing.OneHotEncoder(sparse = False)
X = atributos.fit_transform(dataset.datos[:, :-1])
Y = dataset.datos[:, -1]

clasificador = MultinomialNB(alpha = 1, class_prior = None, fit_prior = True) # Alpha 1 -> Laplace
clasificador.fit(X, Y)

valCruzada = cross_val_score(clasificador, X, Y, cv = validacionCruzada)

print("Validacion Cruzada usando Laplace:")
print("\tMedia: " + str(1 - np.mean(valCruzada)))
print("\tVarianza: " + str(np.var(valCruzada)) + "\n")


Validacion Simple sin usar Laplace:
	Media: 0.2925899999999999
	Varianza: 0.0007733418999999994



Validacion Simple usando Laplace:
	Media: 0.2711800000000001
	Varianza: 0.0008168076000000004



Validacion Cruzada sin usar Laplace:
	Media: 0.29699999999999993
	Varianza: 0.0006559999999999992



Validacion Cruzada usando Laplace:
	Media: 0.275
	Varianza: 0.00025000000000000044



Tenemos los siguientes resultados:

| Dataset       | Tipo Validación        | Laplace | Media de errores | Varianza |
|---------------|------------------------|---------|---------|------------------|
| German        | Simple                 | No      | 0.2925899999999999  | 0.0007733418999999994
| German        | Simple                 | Sí      | 0.2711800000000001  | 0.0008168076000000004
| German        | Cruzada                | No      | 0.29699999999999993 | 0.0006559999999999992
| German        | Cruzada                | Sí      | 0.275 | 0.00025000000000000044

Aquí observamos el efecto que tiene Laplace, particularmente en la validación cruzada, reduciendo la varianza:

Dataset German.csv, Scikit-Learn sin Laplace: 0.297 ± 0.000656  
Dataset German.csv, Scikit-Learn con Laplace: 0.275 ± 0.00025

Sin embargo, apreciamos que esta implementación con la librería SL está cometiendo más errores de media que la nuestra:

Dataset German.csv, Naive-Bayes propio sin Laplace: 0.0015  
Dataset German.csv, Naive-Bayes propio con Laplace: 0.00167

